### DISCUSSION TITLE
Custom Session Data

### DISCUSSION THREAD

**Discussion by fangmarks** [nextauthjs/next-auth#1055] [Help] [answered]
I'm trying to add more Data into the Session when a User logs in. I already tried exposing the data I need during Signup and also tried to edit the data of the JWT as well.

So I'm curious if there's a way to properly add more Data, like `Name`, `username` or `id` into the Session?

**Comment by balazsorban44** [comment, marked as answer]
Please read the documentation first. 

You can use the `jwt` and `session` callbacks to add any data to your session.

https://next-auth.js.org/configuration/callbacks

**Reply by fangmarks** [reply]
I've tried doing that. `token` has all the data I want but session isn't getting any of that data. So now I'm clueless as to how I'd get the data from token into the session, unless you mean that I have to make extra database requests in the session callback just so I can get more data into the session.

![image](https://user-images.githubusercontent.com/20520193/103977794-475bf180-517a-11eb-95ed-f4c3657d0e82.png)

**Reply by fangmarks** [reply]
Got it working by calling my API during the Session callback.

**Reply by balazsorban44** [reply]
You don't need a database to extend your jwt. 

token.user in the jwt callback is only available on the first request though (right after signin). You have to add it in an if statement, otherwise you overwrite the jwt in every subsequent request, removing the user you just added.

Please look at the docs carefully:

https://next-auth.js.org/configuration/callbacks#jwt-callback


Database communication inside the session callback can be expensive, especially since you fetch data that potentially has never changed in the lifetime of the token. The jwt callback is the intended place to extend the jwt token.
